rts/lra/lra-coordinator-jar/src/main/java/io/narayana/lra/coordinator/domain/model/LRARecord.java

```diff
@@ -361,18 +375,89 @@ private int tryDoEnd(boolean compensate) {
                  * so if we were to return FINISH_ERROR recovery would not replay the log.
                  * To force the record to be eligible for recovery we return a heuristic hazard.
                  */
-                return TwoPhaseOutcome.HEURISTIC_HAZARD;
+                return atEnd(TwoPhaseOutcome.HEURISTIC_HAZARD);
             }
 
             status = ParticipantStatus.Completing; // recovery will figure out the status via the status url
 
-            return TwoPhaseOutcome.FINISH_ERROR;
+            return atEnd(TwoPhaseOutcome.FINISH_ERROR);
         }
 
         updateStatus(compensate);
 
         // if the the request is still in progress (ie accepted is true) let recovery finish it
-        return accepted ? TwoPhaseOutcome.HEURISTIC_HAZARD : TwoPhaseOutcome.FINISH_OK;
+        return atEnd(accepted ? TwoPhaseOutcome.HEURISTIC_HAZARD : TwoPhaseOutcome.FINISH_OK);
+    }
+
+    boolean isFinished() {
+        // nested participants must still be able to compensate even if they are closed
+        if (compensateURI == null) {
+            return afterURI != null;
+        }
+
+        switch (status) {
+            case Completed:
+                /* FALLTHRU */
+            case FailedToComplete:
+                return parentId == null; // completed nested LRAs must remain cancellable
+            case Compensated:
+                /* FALLTHRU */
+            case FailedToCompensate:
+                return true;
+            default:
+                return false;
+        }
+    }
+
+    private boolean request(URI target, Transaction lra, String payload) {
+        try {
+            RequestBuilder builder = new RequestBuilder(target)
+                    .request()
+                    .header(LRA.LRA_HTTP_ENDED_CONTEXT_HEADER, lra.getId().toASCIIString())
+                    .header(LRA.LRA_HTTP_RECOVERY_HEADER, recoveryURI.toASCIIString())
+                    .async(PARTICIPANT_TIMEOUT, TimeUnit.SECONDS);
+            ResponseHolder response = target.equals(forgetURI) ? builder.delete() : builder.put(payload, MediaType.TEXT_PLAIN);
+
+            if (response.getStatus() == 200) {
+                return true;
+            }
+        } catch (WebApplicationException e) {
+            if (LRALogger.logger.isInfoEnabled()) {
+                LRALogger.logger.infof("Could not notify URI at %s (%s)", target, e.getMessage());
+            }
+        }
+
+        return false;
+    }
+
+    private int atEnd(int res) {
+        URI nested = compensateURI != null ? extractParentLRA(compensateURI) : null;
```

Comment:
The method understands the narayana LRA format for nested LRAs acting as participants. But since I only need the parent in order to deliver the forget notification and since we don't have a test for that in LRA RC1 (you added it after that in issue 208 / pr 258) I have removed the logic. We can add it back in again when we add an implementation of the MP-LRA test [TckContextTests#testForgetCalledForNestedParticipantsWhenParentIsClosed](https://github.com/eclipse/microprofile-lra/blob/master/tck/src/main/java/org/eclipse/microprofile/lra/tck/TckContextTests.java#L218)

Reply:
it seems that you've removed this line. LRARecord already should have parentId in the field so no need to extract it.

Reply:
... But since I only need the parent in order to deliver the forget notification and since we don't have a test for that in LRA RC1 (you added it after that in issue 208 / pr 258) I have removed the logic. We can add it back in again when we add an implementation of the MP-LRA test TckContextTests#testForgetCalledForNestedParticipantsWhenParentIsClosed